routers/user/auth.go

```diff
@@ -120,6 +127,11 @@ func SignInPost(ctx *context.Context, form auth.SignInForm) {
 	// Clear whatever CSRF has right now, force to generate a new one
 	ctx.SetCookie(setting.CSRFCookieName, "", -1, setting.AppSubUrl)
 
+	if redirectTo := ctx.Query("redirect_to"); len(redirectTo) > 0 {
```

Comment:
What happens is:

 1) URL to a private repository is advertised on a mailing list
 2) Legit users click on the link, with a browser having no login cookie
 3) Legit user get a 404 response, with a "sign-in" link on the top-right corner
 4) Legit user performs the sign-in

Right now the above results in loosing the URL of step 1.
This PR fixes that.